x-pack/plugins/spaces/server/routes/api/external/share_add_spaces.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { schema } from '@kbn/config-schema';
+import { wrapError } from '../../../lib/errors';
+import { ExternalRouteDeps } from '.';
+import { SPACE_ID_REGEX } from '../../../lib/space_schema';
+import { createLicensedRouteHandler } from '../../lib';
+
+const uniq = <T>(arr: T[]): T[] => Array.from(new Set<T>(arr));
+export function initShareAddSpacesApi(deps: ExternalRouteDeps) {
+  const { externalRouter, getSavedObjects } = deps;
+
+  externalRouter.post(
+    {
+      path: '/api/spaces/_share_saved_object_add',
+      validate: {
+        body: schema.object({
+          spaces: schema.arrayOf(
+            schema.string({
+              validate: value => {
+                if (!SPACE_ID_REGEX.test(value)) {
+                  return `lower case, a-z, 0-9, "_", and "-" are allowed`;
+                }
+              },
+            }),
+            {
+              validate: spaceIds => {
+                if (!spaceIds.length) {
+                  return 'must specify one or more space ids';
+                } else if (uniq(spaceIds).length !== spaceIds.length) {
+                  return 'duplicate space ids are not allowed';
+                }
+              },
+            }
+          ),
+          object: schema.object({
+            type: schema.string(),
+            id: schema.string(),
+          }),
+        }),
+      },
+    },
+    createLicensedRouteHandler(async (_context, request, response) => {
+      const { SavedObjectsClient, getScopedSavedObjectsClient } = getSavedObjects();
+      const scopedClient = getScopedSavedObjectsClient(request);
+
+      const spaces = request.body.spaces;
+      const { type, id } = request.body.object;
+
+      try {
+        await scopedClient.addNamespaces(type, id, spaces);
+      } catch (error) {
+        if (SavedObjectsClient.errors.isNotFoundError(error)) {
+          return response.notFound();
```

Comment:
I was basing this behavior off of that in the other space routes, like `delete`, `get`, and `put`. Do you think we should change those too?

Reply:
It's a good question, but I think there's a distinction here: The `delete`, `get` and `put` routes all deal with the management of the `space` saved object types, but the fact that these are saved objects is an implementation detail of the Spaces feature. Surfacing the underlying error thrown by the saved objects client could be misleading in that case, as we'd get an error back like `Saved object [space/default] not found`. Ultimately these existing endpoints end up being just as vague though, because we really should surface an error like `Space [default] not found`, or similar.

These new endpoints explicitly deal with saved objects, and the consumer is aware of that fact, so I think it makes sense to return the underlying saved object errors if we encounter them.

As for updating the existing spaces routes, I think that's a good exercise for https://github.com/elastic/kibana/issues/49334